src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfJsonPrimitive.cs

```diff
@@ -0,0 +1,241 @@
+                    }
+                    else
+                    {
+                        quotedStringValue = new byte[2 + reader.ValueSpan.Length];
+                        reader.ValueSpan.CopyTo(quotedStringValue.AsSpan(1, reader.ValueSpan.Length));
+                    }
+
+                    quotedStringValue[0] = quotedStringValue[quotedStringValue.Length - 1] = JsonConstants.Quote;
+                    return new JsonValueOfJsonString(quotedStringValue, reader.ValueIsEscaped, options);
+                case JsonTokenType.Number:
+                    byte[] numberValue = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan.ToArray();
+                    return new JsonValueOfJsonNumber(new JsonValueOfJsonNumber.JsonNumber(numberValue), options);
+                default:
+                    Debug.Fail("Only primitives allowed.");
+                    ThrowHelper.ThrowJsonException();
+                    return null!; // Unreachable, but required for compilation.
+            }
+        }
+
+        private sealed class JsonValueOfJsonString : JsonValue
+        {
+            private readonly byte[] _value;
+            private readonly bool _isEscaped;
+            private readonly JsonSerializerOptions _serializerOptions;
+
+            internal JsonValueOfJsonString(byte[] value, bool isEscaped, JsonSerializerOptions options)
+                : base(options.GetNodeOptions())
+            {
+                _value = value;
+                _isEscaped = isEscaped;
+                _serializerOptions = options;
+            }
+
+            private ReadOnlySpan<byte> ValueWithoutQuotes => _value.AsSpan(1, _value.Length - 2);
+
+            internal override JsonNode DeepCloneCore() => new JsonValueOfJsonString(_value, _isEscaped, _serializerOptions);
+            private protected override JsonValueKind GetValueKindCore() => JsonValueKind.String;
+
+            public override void WriteTo(Utf8JsonWriter writer, JsonSerializerOptions? options = null)
+            {
+                ArgumentNullException.ThrowIfNull(writer);
+
+                if (_isEscaped)
+                {
+                    writer.WriteStringValue(JsonReaderHelper.GetUnescapedString(ValueWithoutQuotes));
+                }
+                else
+                {
+                    writer.WriteStringValue(ValueWithoutQuotes);
+                }
+            }
+
+            public override T GetValue<T>()
+            {
+                if (!TryGetValue(out T? value))
+                {
+                    ThrowHelper.ThrowInvalidOperationException_NodeUnableToConvertElement(JsonValueKind.String, typeof(T));
+                }
+
+                return value;
+            }
+
+            public override bool TryGetValue<T>([NotNullWhen(true)] out T? value)
+                where T : default
+            {
+                if (!_serializerOptions.TryGetTypeInfo(typeof(T), out JsonTypeInfo? ti))
+                {
```

Comment:
One potential source of regression here might be source generators. If a source generated context doesn't include a contract for, say, `single` then converting the value to `single` would now start to fail. As much as I hate to say this, we might want to keep the original hardcoded conversion checks before falling back to this catch-all.